### ISSUE TITLE
Black video when CMS enabled

### ISSUE THREAD

**Issue by arucard21** [xbmc/xbmc#15183]
### Describe the bug
Here is a clear and concise description of what the problem is:

When I play any video with Color Management enabled in the Settings, the video itself is entirely black. The audio and subtitles still work correctly though. 

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

I expect that the video is shown correctly, with its colors corrected as needed.

## Actual Behavior

The video is entirely black.

## Possible Fix

The problem seems to be with Color Management. I used this in v17 where it worked without problems. The obvious workaround is to disable Color Management. The video is then shown again but its colors will not be corrected.

### To Reproduce

1. In Settings > System > Display, enable Color Management with an ICC profile (rest can be default)
2. Ensure that no hardware acceleration is used (since Color Management doesn't work with that) to play videos
3. Play any video. The video is now shown entirely black.

### Debuglog

The debuglog can be found here: https://paste.kodi.tv/ucuqevilub.kodi
This debug log was created by starting Kodi (with debug logging already enabled), starting a video (to reproduce the bug) and then closing Kodi again.

### Screenshots 

No screenshots since it seems unnecessary to show a screenshot of a black screen.

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

Kodi was compiled using cmake with LCMS2 enabled. This worked correctly in v17/Krypton. I've tried modifying some of the Color Management settings but that didn't make a difference. I could not test this with the 3D LUT option of Color Management since I do not have a 3D LUT file.

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [x] Linux
 - [ ] OSX
 - [ ] Raspberri-Pi
 - [ ] Windows
 - [ ] Windows UWP

 - Debian (unstable with kernel 4.19)
 - Kodi version: 18.0rc4-Leia (git tag)
 - LCMS2 version: 2.9

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by DmnzH** [comment]
I am having the same issue, On Ubuntu with Kodi v18 from ppa:team-xbmc/ppa. Adding a 3dlut file does not correct the issue. Ubuntu 18.04 LTS, Linux kernel 4.15

**Comment by arucard21** [comment]
I just installed Kodi v18.1 (compiled myself on Debian) and the problem still occurs.

**Comment by kib** [comment]
if anyone wants to take a stab at fixing this (@fritsch @laurimyllari ?), i can test and reproduce the issue and have a colorimeter available.

**Comment by haakobja** [comment]
Seems like it's still an issue in Kodi 18.2. I'll post a debug log later

**Comment by arucard21** [comment]
I just checked if this bug is still present in v18.4 and unfortunately it is. 

I went through the debug logs a bit to see if there's any indication of what might be causing this and found the follow bit in the logs. It shows up whenever I play a video with CMS enabled. That's both in the case where you start a video with CMS already enabled as well as when you enable CMS while your video is playing. 
```
2019-09-13 08:37:57.121 T:140417609353920   DEBUG: PrepareNextRender: frameOnScreen: 13284060.044152 renderPts: 13348703.715903 nextFramePts: 13388000.000000 -> diff: -39296.284097  render: 0 forceNext: 0
2019-09-13 08:37:57.121 T:140417609353920 WARNING: CLinuxRendererGL::UpdateVideoFilter - chosen scaling method 13, is not supported by renderer
2019-09-13 08:37:57.122 T:140417609353920   DEBUG: GL: Requested render method: 0
2019-09-13 08:37:57.122 T:140417609353920   DEBUG: GL: BaseYUV2RGBGLSLShader: defines:
                                            #define XBMC_DITHER
                                            #define XBMC_FULLRANGE
                                            #define KODI_3DLUT
                                            #define XBMC_texture_rectangle 0
                                            #define XBMC_texture_rectangle_hack 0
                                            #define XBMC_STRETCH 0
                                            #define XBMC_YV12
2019-09-13 08:37:57.122 T:140417609353920  NOTICE: GL: Selecting YUV 2 RGB shader
2019-09-13 08:37:57.122 T:140417609353920  NOTICE: GL: NPOT texture support detected
2019-09-13 08:37:57.122 T:140417609353920  NOTICE: GL: Using GL_ARB_pixel_buffer_object
2019-09-13 08:37:57.137 T:140417609353920   ERROR: GetDirectory - Error getting
```
I think that maybe this part: `chosen scaling method 13, is not supported by renderer` might have something to do with the problem. Though I don't really know what it means. I tried changing the Video Settings to make them as simple as possible (no deinterlace, nearest neighbor scaling, normal video mode, post processing and tone mapping disabled. I also disabled dithering in the System Settings already) but the video was still black. I hope this might help someone else figure out what's going wrong here.

**Comment by arucard21** [comment]
Just tried this in v18.5 and this problem still occurs. I also converted my ICC profile to a (MadVR) 3DLUT file to test whether that works and it has the same problem. 

Interestingly, when you switch to 3DLUT (from ICC) in Kodi the screen is no longer black if you don't have a (valid) 3DLUT file selected. Only when you select the 3DLUT file, the screen goes black again. Not sure if this means anything when it comes to debugging this but I thought I'd mention it.

**Comment by fritsch** [comment]
Sadly this part of kodi is currently unmaintained as many others. If time allows I will just remove it for 18.6

If a volunteer reads this, now is the time to jump in.

**Comment by arucard21** [comment]
@fritsch I wouldn't want this to get removed, so I'd like to look into this a bit and see if I can fix it (though I'm not too familiar with CMS or how it works in Kodi). I couldn't really find any good documentation about how to debug Kodi (so I can set breakpoints and inspect variables and such, instead of just adding logging lines everywhere). I might be looking in the wrong places though, any tips about where to find such documentation?

**Comment by cpmw167** [comment]
@fritsch Please don't remove CMS. This seems to be quite trivial to fix.

Changing the "format" paramater in glTexImage3D from GL_RGB16 to GL_RGB fixes CMS on my system.

I mean changing this line in xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp
glTexImage3D(GL_TEXTURE_3D, 0, GL_RGB16, m_CLUTsize, m_CLUTsize, m_CLUTsize, 0, GL_RGB16, GL_UNSIGNED_SHORT, m_CLUT);

into this should fix it.
glTexImage3D(GL_TEXTURE_3D, 0, GL_RGB16, m_CLUTsize, m_CLUTsize, m_CLUTsize, 0, GL_RGB, GL_UNSIGNED_SHORT, m_CLUT);

According to Khronos OpenGL registry  "GL_RGB16" is not a valid value for "format" parameter of glTexImage3D.

**Comment by fritsch** [comment]
@cpmw167 Thanks much. I will PR your change. That looks sane.

**Comment by a1rwulf** [comment]
@arucard21 in case you still need to know how to debug kodi: Specifying `-DCMAKE_BUILD_TYPE=Debug` in your cmake command will build kodi with debug symbols.

**Comment by arucard21** [comment]
@cpmw167 I'm really happy you figured that out. I tried looking into this but it was way outside of my expertise so I didn't get too far with it. I'm very happy for this to be working again. 

@a1rwulf Thanks for the help. I did try that and got gdb somewhat working (I think). I think this has more to do with me being unfamiliar with gdb (and C++ debugging in general). But at least it's good to know that I was on the right track with that build type parameter.
